### DISCUSSION TITLE
Socket.io equivalent ?

### DISCUSSION THREAD

**Discussion by seba9999** [geckosio/geckos.io#78] [Help / Questions]
I'm trying to migrate my app from socket.io to geckos.io because I need it to be faster ...

But I'm wondering if those 3 points can be done with geckos.io :

1. Certificates
2. Intercept all packets (using [Middlewares](https://socket.io/docs/v3/middlewares/))
3. Acknowledgement functions

### 1) Certificates
For instance, I'm using ( server side ) socket.io like this for certificates : 

```
const fs = require('fs');
const socketio = require('socket.io');

let https;
if (isProduction) {
  https = require('https');
} else {
  https = require('http');
}

let options = null;
const port = 7776;

if (isProduction) {
  options = {
    key: fs.readFileSync(
      '/etc/letsencrypt/live/myConf.com/privkey.pem'
    ),
    cert: fs.readFileSync(
      '/etc/letsencrypt/live/myConf.com/fullchain.pem'
    )
  };
}

const server = https.createServer(options);
const io = socketio(server, {
  pingInterval: 45000,
  pingTimeout: 600000
});

server.listen(port);
```

### 2) intercept packets 

I mean test if the request is legitimate ... I do something like this with socket.io : 

```
  socket.use((packet, next) => {
    const isAuthorized = checkTheAuthorisationOfThisPacket(packet);
    if (isAuthorized) {
      next();
    } else {
      console.log('unauthorized request : ', packet[0]);
      socket.emit('unAuthorized');
    }
  });
```

### 3) Acknowledgement functions

With socket.io you can send a function client side but execute it server side :

```
  socket.on('test', callbackFromClient => {
    callbackFromClient ();
  });
```

Do all this stuff can be made with geckos.io ?

**Comment by yandeu** [comment]
1. Yes, it is the same. See [servers](https://github.com/geckosio/geckos.io#servers).
2. You can use the [Authorization header](https://github.com/geckosio/geckos.io#authorization-and-authentication).
3. I've never seen a functionality like this.

**Reply by seba9999** [reply]
Thanks for your reply I should try all of this ...

Regarding the acknowledge function I think that I've miss-explained what it is ... It doesn't "execute" server side ... It's just a callback definition send to the server and the server call that function.

Please see [Socket.io docs](https://socket.io/docs/v3/emitting-events/#Acknowledgements)

**Reply by jdspugh** [reply]
I think you're describing RPC. You can build your own RPC on top of Geckos if you need it. It's not required for multiplayer games in general.

**Comment by mowses** [comment]
I would join this discussion because I actually need the #2 (intercept Packets). My app needs constant query over a graphQL api which needs some headers to be sent from the client to the server (cookies, x-xsrf-token).
I could manage the authorization phase because the callback retrieve the request object.
Now I am stuck with subsequent calls which needs the server to call the graphQL api and need these headers again. I believe that a middleware like socket.io does would be suffice to solve my issue.
